Add fees to tx-wide caps

jackcmay opened the pull request:
#### Problem

Tx-wide compute unit caps default to 200k and reject all transactions that don't request more.  That will immediately break lots of clients.

#### Summary of Changes

- Default to the max compute unit cap for transactions that don't explicitly request a cap
- Charge a fee based on the cap requested or assigned to the transaction.
- Change the max compute unit cap to be 1.4m units

TODO: 
- Need feedback on the fee changes.  Fees are defined as a lamports_per_signature charged for each signature and a series of bins of compute unit => fee.  Using bins allows us to apply whatever fee structure/curve is appropriate.

Fixes #

jackcmay commented:
@aeyakovenko @sakridge Please review and I'm looking for input on what the fee schedule should be based on the requested compute units.  We can also expand to include a fee based on both signatures and requested compute units.

jackcmay commented:
@t-nelson When this feature is activated, all nonce'd fee's will also switch to the new schedule, the nonce'd value will hold the congestion based multiplier (like it does now) (which pretty much never changes)

jackcmay commented on sdk/src/fee.rs:
```diff
@@ -0,0 +1,60 @@
+            .saturating_add(self.bins.last().map(|bin| bin.fee).unwrap_or_default())
+    }
+}
+impl Default for FeeStructure {
```
@aeyakovenko @sakridge 

The fee structure consists of a series of bins (cap, fee), when transactions come in they are charged based on which bin their requested compute units lands, and how many signatures their transaction contains.  If they don't request they get the highest bin.  The bins are currently a x10 lower fee for < 10k units, the current fee for 10k-200k units, and a doubling after that.  Can you provide feedback on these levels?  The activation of this feature will be via governance and discussed further beforehand but I'd like to get this initial shot in the ballpark.  I'm also working on a mechanism to which we can modify fee values at epoch boundaries without making a new release, the authority to do that will also be managed by a governance program.

aeyakovenko reviewed (commented):
can ComputeBudgetInstruction::request_units take more than one arg?  or ComputeBudgetInstruction::request_units_with_fee(additional_fee: Lamports, units)

I think shortest path for us to deal with spam prioritization when building blocks is to sort by "additional_fee/total CU's". then we don't need the bins or a static fee for CUs

aeyakovenko commented:
I think the only complexity with "additonal_fee" arg would be in the wallet/UI.  we would need to give an estimate based on the lowest CU rate in the block and the worst case for the lowest CU rate for any writable accounts that are saturated.

aeyakovenko commented:
I think the only complexity with "additonal_fee" arg would be in the wallet/UI.  we would need to give an estimate based on the lowest CU rate in the block and the worst case for the lowest CU rate for any writable accounts that are saturated.

aeyakovenko commented:
@taozhu-chicago, @sakridge if we add `ComputeBudgetInstruction::request_units_with_fee(additional_fee: Lamports, units)` an additiona_fee parameter, it should be easy to sort by `additional_fee/total tx cu's` after sigverify and before forwarding.  wdyt?

sakridge commented:
> @taozhu-chicago, @sakridge if `ComputeBudgetInstruction::request_units_with_fee(additional_fee: Lamports, units)` has an additional fee parameter, it should be easy to sort by `additional_fee/total tx cu's` after sigverify and before forwarding. wdyt?
> 
> Write accounts would be saturated by the highest paid compute units first. I think its probably sub optimal, since the block would earn more rewards if it packed reads first, but this seems pretty easy to ship and would unblock users when a market is saturated.

yea, I like it.

aeyakovenko commented:
@sakridge @taozhu-chicago we would need to add something to bench-tps that validates that higher paying txs go through with high reliability.

aeyakovenko commented:
@sakridge @jackcmay the fee parameter doesn't preclude us from shipping #21883 later.  it just seems like a way more complex change since it requires a rebate.

taozhu-chicago commented:
> > @taozhu-chicago, @sakridge if `ComputeBudgetInstruction::request_units_with_fee(additional_fee: Lamports, units)` has an additional fee parameter, it should be easy to sort by `additional_fee/total tx cu's` after sigverify and before forwarding. wdyt?
> > Write accounts would be saturated by the highest paid compute units first. I think its probably sub optimal, since the block would earn more rewards if it packed reads first, but this seems pretty easy to ship and would unblock users when a market is saturated.
> 
> yea, I like it.

To clarify, the `additional_fee` is the fee user is willing to pay on top of CUs (so spam TXs would have low value in this param), or is it required fee for writing to accounts?

It's good to have a way to prioritize TXs by its requested budget, more precise than prioritize based on sender stake weight. Should they co-exist? if so, would you think this should happen before or after Stake Weight Prioritizing?

aeyakovenko commented:
> To clarify, the `additional_fee` is the fee user is willing to pay on top of CUs (so spam TXs would have low value in this param), or is it required fee for writing to accounts?

yea, its what the user is willing to pay on top of the base fee.

> It's good to have a way to prioritize TXs by its requested budget, more precise than prioritize based on sender stake weight. Should they co-exist? if so, would you think this should happen before or after Stake Weight Prioritizing?

I think stake weight is how the queue is filled, but once its filled its prioritized in by 'fee/compute' and fills accounts in that order.  We need the forwarders and RPCs to estimate how accounts are filled by the leader and not just saturate the queue prior to the bank with a single writable account and all the high paying txs there.  otherwise blocks won't be filled.

jackcmay commented:
@aeyakovenko I changed this PR:
- added the "additional_fee" to the compute request instruction
- Updated the fee calculations to base fee on sigs, write-locks, additional fee, and compute unit base fee
- Set the default fee values to match the current fee structure, so values effectively result in per-sig fee
- Planning to enable this in v1.9 with our without compute related fees
- Follow up with governance based fee value setting changes
